src/py/flwr/common/backoff.py

```diff
@@ -0,0 +1,278 @@
+    Iterable,
+    Optional,
+    Tuple,
+    Type,
+    Union,
+)
+
+
+def exponential(
+    base_delay: float = 1,
+    multiplier: float = 2,
+    max_delay: Optional[int] = None,
+) -> Generator[float, None, None]:
+    """Wait time generator for exponential backoff strategy.
+
+    Parameters
+    ----------
+    base_delay: float (default: 1)
+        Initial delay duration before the first retry.
+    multiplier: float (default: 2)
+        Factor by which the delay is multiplied after each retry.
+    max_delay: Optional[float] (default: None)
+        The maximum delay duration between two consecutive retries.
+    """
+    delay = base_delay if max_delay is None else min(base_delay, max_delay)
+    while True:
+        yield delay
+        delay *= multiplier
+        if max_delay is not None:
+            delay = min(delay, max_delay)
+
+
+def constant(
+    interval: Union[float, Iterable[float]] = 1,
+) -> Generator[float, None, None]:
+    """Wait time generator for specified intervals.
+
+    Parameters
+    ----------
+    interval: Union[float, Iterable[float]] (default: 1)
+        A constant value to yield or an iterable of such values.
+    """
+    if not isinstance(interval, Iterable):
+        interval = itertools.repeat(interval)
+    yield from interval
+
+
+def random_jitter(value: float) -> float:
+    """Jitter the value by a random number of milliseconds.
+
+    This adds up to 1 second of additional time to the original value.
+    Prior to backoff version 1.2, this was the default jitter behavior.
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return value + random.random()
+
+
+def full_jitter(value: float) -> float:
+    """Jitter the value across the full range (0 to value).
+
+    This corresponds to the "Full Jitter" algorithm specified in the
+    AWS blog post on the performance of various jitter algorithms.
+    See: http://www.awsarchitectureblog.com/2015/03/backoff.html
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return random.uniform(0, value)
+
+
+@dataclass
+class Details:
+    """Details for event handlers in SafeInvoker."""
+
+    func: Callable[[Any], Any]
+    args: Tuple[Any, ...]
+    kwargs: Dict[str, Any]
+    tries: int
+    elapsed: float
+    exception: Optional[Exception] = None
+    wait: Optional[float] = None
+
+
+# pylint: disable-next=too-many-instance-attributes
+class SafeInvoker:
+    """Wrapper class for backoff and retry triggered by exceptions.
+
+    Parameters
+    ----------
+    wait: Generator[float, None, None]
+        A generator yielding successive wait times in seconds. If the generator
+        is finite, the giveup event will be triggered when the generator raises
+        `StopIteration`.
+    exception:  Union[Type[Exception], Iterable[Type[Exception]]]
+        An exception type (or iterable of types) that triggers backoff.
```

Comment:
Wouldn't it be sufficient to support the `Iterable` version? The version with just one `Exception` is still possible via a single-element list.

Reply:
I believe it would be better to accept both versions. One example is the built-in `isinstance` function. It would actually be a bit "uncomfortable" to write `isinstance(obj, [MyClass])` comparing to `isinstance(obj, MyClass)`.

Reply:
From this [SO answer](https://stackoverflow.com/a/6470452), shouldn't the type be just `Tuple[Exception]`?

Reply:
Just double checked. Yes, the type should be `Tuple[Exception, ...]` instead of any iterable. I've changed the type hint.